### PR TITLE
fix/csv queuer fetcher

### PR DESCRIPTION
**Pull request by thepsalmist**
The PR builds from #271 to fetch urls for CSV files on S3 for Database E dates 01/25-02/17

### CODE REVIEW

**Comment by philbudne** [comment]
Some preliminary comments:
1. The branch seems to contain other stuff (`HIST_YEAR=2023` and `add ELASTICSEARCH_SNAPSHOT_REPO`)?  It's best practice to have PRs based on main.
2. I think the pipeline type "csv-fetcher" is misleading; I suggest using "csv": It runs a different queuer that feeds into the (regular) queue-based fetcher and I think it's likely to be (re)used in the future, as opposed to being used only once.
3. In csv-queuer:
  * I don't think any of the commented out code in csv-queuer.py needs to be kept around. 
  * the comment `# let hist-fetcher quarantine if bad` should go away
  * the large block comment on `urls_seen` is no longer meaningful (since the CSV file is unlikely to have come from the legacy system), BUT filtering for duplicate urls is probably still a good idea (doesn't cost much, and can save time/effort).
 * the comment block `# content_metadata.parsed_date is not set, so parser.py will` can go away 
 *  `rss.source_feed_id = None` and `rss.source_source_id = None` can go away (should be the default value)
  * `rss.source_url = url` can go away: we don't have the URL of the RSS file the story was found in.
4. in deploy.sh:
 * `PIPE_TYPE_PFX='hist-'` I suggest use `csv-` and setting `PORT_BIAS=600` to allow co-existence with other stacks
 * Would be good to set "ARCH_SUFFIX=csv" so the generated WARC files are distinct from the current day ones

File: indexer/workers/fetcher/csv-queuer.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        urls_seen: Set[str] = set()
+
+        # only url column
+        for row in csv.reader(io.TextIOWrapper(fobj)):
+            logger.debug("%r", row)
+
+            url = row[0]
+            if not isinstance(url, str) or not url:
+                self.incr_stories("bad-url", repr(url))
+                continue
```
Comment (kilemensi) [inline indexer/workers/fetcher/csv-queuer.py:47]: Aside (not for this PR): Seems like `self.check_story_url` does most of the check that this if does and we _probably_ don't need both (similarly in `hist-queue.py`)

File: docker/docker-compose.yml.j2 [hunk 2]
```diff
@@ -328,7 +328,7 @@ services:
       QUEUER_S3_ACCESS_KEY_ID: {{queuer_s3_access_key_id}}
       QUEUER_S3_REGION: {{queuer_s3_region}}
       QUEUER_S3_SECRET_ACCESS_KEY: {{queuer_s3_secret_access_key}}
-{% elif pipeline_type == 'queue-fetcher' %}
+{% elif pipeline_type == 'queue-fetcher' or pipeline_type == 'csv' %}
```
[the remark this answers is not in the document]
Comment (thepsalmist) [reply docker/docker-compose.yml.j2:331]: We still want to spin the `queue-fetcher`, to do the fetching although being fed from a `csv-queuer` not the typical `rss-queuer`

File: indexer/workers/fetcher/csv-queuer.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+            if not self.check_story_url(url):
+                continue  # logged and counted
+
+            story = Story()
+            with story.rss_entry() as rss:
+                rss.link = url
+                rss.fetch_date = fetch_date
+                rss.via = fname
+                rss.source_url = url
```
[the remark this answers is not in the document]
Comment (thepsalmist) [reply indexer/workers/fetcher/csv-queuer.py:61]: > Do we know this?

Could you clarify on this?

File: indexer/workers/fetcher/csv-queuer.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+            if url in urls_seen:
+                self.incr_stories("dups", url)
+                continue
+
+            if not self.check_story_url(url):
+                continue  # logged and counted
+
+            story = Story()
+            with story.rss_entry() as rss:
```
[the remark this answers is not in the document]
Comment (thepsalmist) [reply indexer/workers/fetcher/csv-queuer.py:57]: Sure

File: indexer/workers/fetcher/csv-queuer.py (hunk 1 above)
Comment (thepsalmist) [reply indexer/workers/fetcher/csv-queuer.py:47]: Yes, `self.check_story_url` does most of the validation, one edge case not handled when url = "".

File: indexer/workers/fetcher/csv-queuer.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+            self.send_story(story)  # calls incr_story: to increment and log
+            urls_seen.add(url)  # mark URL as seen
+
+
+if __name__ == "__main__":
+    run(
+        CSVQueuer,
+        "csv-queuer",
+        "Read CSV of historical stories, queue to hist-fetcher",
```
Comment (philbudne) [inline indexer/workers/fetcher/csv-queuer.py:75]: "queue to regular fetcher"

**Review by philbudne — CHANGES_REQUESTED** [review]

File: indexer/workers/fetcher/csv-queuer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kilemensi) [reply indexer/workers/fetcher/csv-queuer.py:62]: Sorry... Been a while but _if I remember correctly_, `rss.source_url` the URL of the RSS file itself. Hence, do we know if `url` is `rss.source_url`?

File: indexer/workers/fetcher/csv-queuer.py (hunk 1 above)
Comment (kilemensi) [reply indexer/workers/fetcher/csv-queuer.py:47]: [`if not url:`](https://github.com/mediacloud/story-indexer/blob/3b10677cf4c2e7ef24306e1a5c4993418821f06c/indexer/storyapp.py#L122) should handle when `url` is `""`.

**Review by philbudne — APPROVED** [review]

**Pull request merged**